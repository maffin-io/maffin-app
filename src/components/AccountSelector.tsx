import React from 'react';
import { useHotkeys } from 'react-hotkeys-hook';
import Select, {
  SingleValue,
  components,
} from 'react-select';
import { BiSearch } from 'react-icons/bi';
import useSWRImmutable from 'swr/immutable';

import { Account } from '@/book/entities';
import { getAccountsWithPath } from '@/book/queries';

export type AccountSelectorProps = {
  id?: string,
  placeholder?: string,
  onChange?: Function,
  showRoot?: boolean,
  isClearable?: boolean,
  ignoreAccounts?: string[],
  defaultValue?: Account,
  className?: string,
};

export default function AccountSelector(
  {
    placeholder,
    onChange,
    ignoreAccounts,
    id = 'accountSelector',
    showRoot = false,
    isClearable = true,
    defaultValue,
    className = '',
  }: AccountSelectorProps,
): JSX.Element {
  const ref = React.useRef<HTMLDivElement>();
  useHotkeys('meta+k', () => {
    ref.current?.focus();
  });

  let { data: accounts } = useSWRImmutable(
    '/api/accounts',
    getAccountsWithPath,
  );

  accounts = accounts || [];
  accounts = accounts.filter(account => !(ignoreAccounts || []).includes(account.type));
  if (!showRoot) {
    accounts = accounts.filter(account => account.type !== 'ROOT');
  }

  return (
    <Select
      // @ts-ignore
      ref={ref}
      id={id}
      name={id}
      aria-label={id}
      options={accounts}
      placeholder={placeholder || 'Choose account'}
      onChange={(newValue: SingleValue<Account> | null) => {
        if (onChange) {
          onChange(newValue);
        }
      }}
      isClearable={isClearable}
      defaultValue={defaultValue}
      backspaceRemovesValue
      onKeyDown={(e: React.KeyboardEvent) => {
        if (e.key === 'Escape') {
          ref.current?.blur();
        }
      }}
      getOptionLabel={(option) => option.path}
      getOptionValue={(option) => option.path}
      openMenuOnFocus
      components={{ Control }}
      className={className}
      styles={{
        option: () => ({}),
      }}
      classNames={{
        // We set h-9 so it's the same as other inputs
        control: () => '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none',
        indicatorSeparator: () => 'hidden',
        menu: () => 'text-sm !bg-gunmetal-800',
        option: (state) => {
          const styles = 'w-full cursor-default hover:bg-gunmetal-700 px-4 py-2';
          if (state.isFocused) {
            // For some reason putting focus:bg-gunmetal-700 doesnt work
            return `bg-gunmetal-700 ${styles}`;
          }
          return styles;
        },
        singleValue: () => '!text-inherit',
        input: () => '!text-inherit',
      }}
    />
  );
}

function Control(
  {
    children,
    ...props
  }: any,
): JSX.Element {
  return (
    <components.Control {...props}>
      <BiSearch className="absolute text-md left-[10px]" />
      {children}
    </components.Control>
  );
}
