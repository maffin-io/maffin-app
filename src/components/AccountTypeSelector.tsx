import React from 'react';
import Select, {
  SingleValue,
  components,
} from 'react-select';
import { BiShuffle } from 'react-icons/bi';

import { Account } from '@/book/entities';

export type AccountTypeSelectorProps = {
  id?: string,
  placeholder?: string,
  onChange?: Function,
  disabled?: boolean,
  ignoreTypes?: string[],
  className?: string,
};

export default function AccountTypeSelector(
  {
    placeholder,
    onChange,
    ignoreTypes = [],
    id = 'accountSelector',
    disabled = false,
    className = '',
  }: AccountTypeSelectorProps,
): JSX.Element {
  const ref = React.useRef<HTMLDivElement>();
  const types = Account.TYPES.slice(1).filter(
    type => !ignoreTypes.includes(type),
  ).map(type => ({ type }));

  return (
    <Select
      // @ts-ignore
      ref={ref}
      id={id}
      name={id}
      aria-label={id}
      options={types}
      isDisabled={disabled}
      placeholder={placeholder || 'Choose account type'}
      onChange={(newValue: SingleValue<{ type: string }> | null) => {
        if (onChange) {
          onChange(newValue?.type);
        }
      }}
      isClearable
      backspaceRemovesValue
      onKeyDown={(e: React.KeyboardEvent) => {
        if (e.key === 'Escape') {
          ref.current?.blur();
        }
      }}
      getOptionLabel={(option) => option.type}
      getOptionValue={(option) => option.type}
      openMenuOnFocus
      components={{ Control }}
      className={className}
      styles={{
        option: () => ({}),
      }}
      classNames={{
        // We set h-9 so it's the same as other inputs
        control: ({ isDisabled }) => {
          if (isDisabled) {
            return '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-700 !border-none';
          }
          return '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none';
        },
        indicatorSeparator: () => 'hidden',
        menu: () => 'text-sm !bg-gunmetal-800',
        option: (state) => {
          const styles = 'w-full cursor-default hover:bg-gunmetal-700 px-4 py-2';
          if (state.isFocused) {
            // For some reason putting focus:bg-gunmetal-700 doesnt work
            return `bg-gunmetal-700 ${styles}`;
          }
          return styles;
        },
        singleValue: () => '!text-inherit',
        input: () => '!text-inherit',
      }}
    />
  );
}

function Control(
  {
    children,
    ...props
  }: any,
): JSX.Element {
  return (
    <components.Control {...props}>
      <BiShuffle className="absolute text-md left-[10px]" />
      {children}
    </components.Control>
  );
}
