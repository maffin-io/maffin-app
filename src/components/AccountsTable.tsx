import React from 'react';
import { ColumnDef } from '@tanstack/react-table';
import Link from 'next/link';
import { BiCircle, BiSolidRightArrow, BiSolidDownArrow } from 'react-icons/bi';
import classNames from 'classnames';

import Table from '@/components/Table';
import type { AccountsTree } from '@/types/accounts';

export type AccountsTableProps = {
  tree: AccountsTree,
};

export default function AccountsTable(
  {
    tree,
  }: AccountsTableProps,
): JSX.Element {
  return (
    <Table<AccountsTree>
      columns={columns}
      data={tree.children}
      initialSort={{ id: 'total', desc: true }}
      showHeader={false}
      showPagination={false}
      tdClassName="p-2"
      getSubRows={row => row.children}
    />
  );
}

const columns: ColumnDef<AccountsTree>[] = [
  {
    accessorKey: 'name',
    id: 'name',
    enableSorting: false,
    header: '',
    cell: ({ row }) => (
      <div
        className={`flex pl-${row.depth * 4}`}
      >
        {row.getCanExpand() ? (
          <button
            type="button"
            onClick={() => row.toggleExpanded()}
          >
            {(
              row.getIsExpanded()
                ? <BiSolidDownArrow className="mr-1 text-xs opacity-20" />
                : <BiSolidRightArrow className="mr-1 text-xs opacity-20" />
            )}
          </button>
        ) : (
          <button
            type="button"
            className="cursor-default"
          >
            <BiCircle className="mr-1 text-xs opacity-20" />
          </button>
        )}
        <Link
          className={classNames('badge hover:text-slate-300', {
            'bg-green-500/20 text-green-300': row.original.account.type === 'INCOME',
            'bg-red-500/20 text-red-300': row.original.account.type === 'EXPENSE',
            'bg-cyan-500/20 text-cyan-300': ['ASSET', 'BANK'].includes(row.original.account.type),
            'bg-orange-500/20 text-orange-300': row.original.account.type === 'LIABILITY',
            'bg-violet-500/20 text-violet-300': ['STOCK', 'MUTUAL'].includes(row.original.account.type),
          })}
          href={`/dashboard/accounts/${row.original.account.guid}`}
        >
          {row.original.account.name}
        </Link>
      </div>
    ),
  },
  {
    accessorFn: (row: AccountsTree) => row.total.toNumber(),
    id: 'total',
    header: '',
    cell: ({ row }) => (
      <span>
        {row.original.total.format()}
      </span>
    ),
  },
];
