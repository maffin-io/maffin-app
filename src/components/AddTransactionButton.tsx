import React from 'react';
import { BiPlusCircle } from 'react-icons/bi';

import TransactionForm from '@/components/forms/transaction/TransactionForm';
import { Account } from '@/book/entities';
import Modal from '@/components/Modal';
import Tooltip from '@/components/Tooltip';

type AddTransactionButtonProps = {
  account: Account,
};

export default function AddTransactionButton(
  {
    account,
  }: AddTransactionButtonProps,
): JSX.Element {
  const [isModalOpen, setIsModalOpen] = React.useState<boolean>(false);
  const disabled = ['INCOME', 'EXPENSE'].includes(account.type);

  return (
    <>
      <Modal
        title={`${account.path} transaction`}
        open={isModalOpen}
        setOpen={setIsModalOpen}
      >
        <TransactionForm
          onSave={() => { setIsModalOpen(false); }}
          account={account}
        />
      </Modal>
      <Tooltip
        text="Add transactions from accounts that are not INCOME/EXPENSE/STOCK/MUTUAL"
        show={disabled}
      >
        <button
          type="button"
          className="btn-primary"
          disabled={disabled}
          onClick={() => setIsModalOpen(!isModalOpen)}
        >
          <BiPlusCircle className="mr-1" />
          Add Transaction
        </button>
      </Tooltip>
    </>
  );
}
