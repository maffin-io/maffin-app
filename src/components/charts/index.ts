export { default as Bar } from './Bar';
export { default as Line } from './Line';
export { default as Tree } from './Tree';
export { default as Pie } from './Pie';

export { default as NetWorthHistogram } from './NetWorthHistogram';
