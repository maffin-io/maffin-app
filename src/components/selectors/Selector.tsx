import React from 'react';
import { useHotkeys } from 'react-hotkeys-hook';
import Select, { SingleValue, components } from 'react-select';
import { BiSearch } from 'react-icons/bi';

export type SelectorProps<T> = {
  labelAttribute: string,
  options: T[],
  id?: string,
  placeholder?: string,
  onChange?: Function,
  isClearable?: boolean,
  defaultValue?: T,
  className?: string,
  disabled?: boolean,
};

export default function Selector<T extends object = {}>(
  {
    labelAttribute,
    options,
    placeholder,
    onChange,
    defaultValue,
    id = 'selector',
    isClearable = true,
    className = '',
    disabled = false,
  }: SelectorProps<T>,
): JSX.Element {
  const ref = React.useRef<HTMLDivElement>();
  useHotkeys('meta+k', () => {
    ref.current?.focus();
  });

  return (
    <Select
      // @ts-ignore
      ref={ref}
      id={id}
      name={id}
      aria-label={id}
      options={options}
      placeholder={placeholder || 'Choose an option'}
      onChange={(newValue: SingleValue<T> | null) => {
        if (onChange) {
          onChange(newValue);
        }
        ref.current?.blur();
      }}
      isClearable={isClearable}
      isDisabled={disabled}
      defaultValue={defaultValue}
      backspaceRemovesValue
      onKeyDown={(e: React.KeyboardEvent) => {
        if (e.key === 'Escape') {
          ref.current?.blur();
        }
      }}
      // @ts-ignore dunno how to set proper types here
      getOptionLabel={(option: T) => option[labelAttribute]}
      // @ts-ignore dunno how to set proper types here
      getOptionValue={(option: T) => option[labelAttribute]}
      openMenuOnFocus
      components={{ Control }}
      className={className}
      styles={{
        option: () => ({}),
      }}
      classNames={{
        // We set h-9 so it's the same as other inputs
        control: (state) => {
          const styles = '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none';
          if (state.isDisabled) {
            // For some reason putting focus:bg-gunmetal-700 doesnt work
            return `opacity-50 ${styles}`;
          }
          return styles;
        },
        indicatorSeparator: () => 'hidden',
        menu: () => 'text-sm !bg-gunmetal-800',
        option: (state) => {
          const styles = 'w-full cursor-default hover:bg-gunmetal-700 px-4 py-2';
          if (state.isFocused) {
            // For some reason putting focus:bg-gunmetal-700 doesnt work
            return `bg-gunmetal-700 ${styles}`;
          }
          return styles;
        },
        singleValue: () => '!text-inherit',
        input: () => '!text-inherit',
      }}
    />
  );
}

function Control(
  {
    children,
    ...props
  }: any,
): JSX.Element {
  return (
    <components.Control {...props}>
      <BiSearch className="absolute text-md left-[10px]" />
      {children}
    </components.Control>
  );
}
