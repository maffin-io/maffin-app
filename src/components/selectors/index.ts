export { default as AccountSelector } from './AccountSelector';
export { default as CommoditySelector } from './CommoditySelector';
export { default as AccountTypeSelector } from './AccountTypeSelector';
export { default as NamespaceSelector } from './NamespaceSelector';

export type { AccountSelectorProps } from './AccountSelector';
export type { CommoditySelectorProps } from './CommoditySelector';
export type { AccountTypeSelectorProps } from './AccountTypeSelector';
