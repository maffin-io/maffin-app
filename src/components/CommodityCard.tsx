import React from 'react';
import Link from 'next/link';
import {
  BiLineChart,
  BiMoney,
  BiSolidGhost,
} from 'react-icons/bi';

import { Tooltip, UpgradeTooltip } from '@/components/tooltips';
import { useCommodity, useMainCurrency, usePrices } from '@/hooks/api';
import Loading from '@/components/Loading';
import { IS_PAID_PLAN } from '@/helpers/env';

export type CommodityCardProps = {
  guid: string;
};

export default function CommodityCard({
  guid,
}: CommodityCardProps): JSX.Element {
  const { data: commodity } = useCommodity(guid);
  const { data: mainCurrency } = useMainCurrency();
  const { data: prices } = usePrices({ from: commodity });

  if (!commodity || !prices) {
    return <Loading />;
  }

  let iconComponent = <BiLineChart className="mr-1" />;
  let tooltipContent = (
    <div>
      <p>
        No prices found for this commodity! This means that accounts
        using this commodity don&apos;t have an exchange rate to convert
        to your main currency and we will convert 1 to 1 for now.
      </p>

      <p className="mt-1">
        To fix this, make sure the symbol can be found in

        {' '}
        <Link
          href={`https://finance.yahoo.com/quote/${commodity.mnemonic}`}
          target="blank"
        >
          Yahoo
        </Link>
        . If not, consider adding this commodity
        as &quot;OTHER&quot; type instead and add prices manually.
      </p>
    </div>
  );

  if (commodity.namespace === 'CURRENCY') {
    iconComponent = <BiMoney className="mr-1" />;
    tooltipContent = (
      <div>
        <p>
          No prices found for the
          {' '}
          {`${commodity.mnemonic}.${mainCurrency?.mnemonic}`}
          {' '}
          pair! This means that accounts using this currency don&apos;
          have an exchange rate to convert to your main currency and we
          will convert 1 to 1 for now.
        </p>

        <p className="mt-1">
          To fix this, make sure the currency can be found in
          {' '}
          <Link
            href={`https://finance.yahoo.com/quote/${commodity.mnemonic}${mainCurrency?.mnemonic}=X`}
            target="blank"
          >
            Yahoo
          </Link>
          . If it does not exist, it means you will have to track prices for this currency
          manually.
        </p>
      </div>
    );
  } else if (commodity.namespace === 'OTHER') {
    tooltipContent = (
      <div>
        <p>
          No prices found for this commodity! This means that accounts
          using this commodity don&apos;t have an exchange rate to convert
          to your main currency and we will convert 1 to 1 for now.
        </p>

        <p className="mt-1">
          To fix this, make sure to add prices manually for this commodity
        </p>
      </div>
    );
    iconComponent = <BiSolidGhost className="mr-1" />;
  }

  let priceAvailable = false;
  if (commodity.namespace === 'CURRENCY') {
    const price = prices.getPrice(commodity.mnemonic, mainCurrency?.mnemonic || '');
    priceAvailable = price?.guid !== 'missing_price';
  } else {
    const price = prices.getInvestmentPrice(commodity.mnemonic);
    priceAvailable = price?.guid !== 'missing_price';
  }

  return (
    <>
      <Link
        key={commodity.guid}
        href={`/dashboard/commodities/${commodity.guid}`}
        className="card col-span-2 cursor-pointer hover:shadow-lg text-slate-400 dark:hover:text-white"
      >
        <div className="flex items-center">
          {iconComponent}
          {commodity.mnemonic}
          {
            !priceAvailable
            && (
              <div
                className="badge warning ml-auto"
                data-tooltip-id={`missing-price-${guid}`}
              >
                !
              </div>
            )
          }
        </div>
      </Link>

      {
        !priceAvailable
        && (
          IS_PAID_PLAN
            ? (
              <Tooltip
                id={`missing-price-${guid}`}
                className="!w-1/6 !text-xs"
                clickable
              >
                {tooltipContent}
              </Tooltip>
            )
            : (
              <UpgradeTooltip
                id={`missing-price-${guid}`}
                message="Non paid users need to add prices manually. Consider upgrading for getting a bunch of commodities synced automatically for you!"
              />
            )
        )
      }
    </>
  );
}
