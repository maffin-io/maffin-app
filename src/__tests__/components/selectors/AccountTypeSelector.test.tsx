import React from 'react';
import { render, screen } from '@testing-library/react';
import { SWRConfig } from 'swr';

import Selector from '@/components/selectors/Selector';
import { AccountTypeSelector } from '@/components/selectors';

jest.mock('@/components/selectors/Selector', () => jest.fn(
  () => <div data-testid="Selector" />,
));
const SelectorMock = Selector as jest.MockedFunction<typeof Selector>;

describe('AccountTypeSelector', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected', async () => {
    const { container } = render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <AccountTypeSelector />
      </SWRConfig>,
    );

    await screen.findByTestId('Selector');
    expect(Selector).toHaveBeenCalledWith(
      {
        id: 'typeSelector',
        isClearable: true,
        defaultValue: undefined,
        className: '',
        disabled: false,
        labelAttribute: 'type',
        options: [
          { type: 'ASSET' },
          { type: 'BANK' },
          { type: 'CASH' },
          { type: 'EQUITY' },
          { type: 'LIABILITY' },
          { type: 'INCOME' },
          { type: 'EXPENSE' },
          { type: 'MUTUAL' },
          { type: 'STOCK' },
        ],
        placeholder: 'Choose account type',
        onChange: expect.any(Function),
      },
      {},
    );
    expect(container).toMatchSnapshot();
  });

  it('passes data as expected', async () => {
    const mockOnChange = jest.fn();
    const { container } = render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <AccountTypeSelector
          id="customId"
          placeholder="My placeholder"
          isClearable={false}
          disabled
          className="class"
          defaultValue={{ type: 'ASSET' }}
          onChange={mockOnChange}
        />
      </SWRConfig>,
    );

    await screen.findByTestId('Selector');
    expect(Selector).toHaveBeenCalledWith(
      {
        id: 'customId',
        isClearable: false,
        disabled: true,
        defaultValue: { type: 'ASSET' },
        className: 'class',
        labelAttribute: 'type',
        options: [
          { type: 'ASSET' },
          { type: 'BANK' },
          { type: 'CASH' },
          { type: 'EQUITY' },
          { type: 'LIABILITY' },
          { type: 'INCOME' },
          { type: 'EXPENSE' },
          { type: 'MUTUAL' },
          { type: 'STOCK' },
        ],
        placeholder: 'My placeholder',
        onChange: expect.any(Function),
      },
      {},
    );
    const { onChange } = SelectorMock.mock.calls[0][0];
    if (onChange) {
      onChange({ type: 'ASSET' });
    }
    expect(mockOnChange).toHaveBeenCalledWith('ASSET');
    expect(container).toMatchSnapshot();
  });

  it('filters types', async () => {
    render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <AccountTypeSelector
          ignoreTypes={['ASSET', 'BANK']}
        />
      </SWRConfig>,
    );

    await screen.findByTestId('Selector');
    expect(Selector).toHaveBeenCalledWith(
      expect.objectContaining({
        options: [
          { type: 'CASH' },
          { type: 'EQUITY' },
          { type: 'LIABILITY' },
          { type: 'INCOME' },
          { type: 'EXPENSE' },
          { type: 'MUTUAL' },
          { type: 'STOCK' },
        ],
      }),
      {},
    );
  });
});
