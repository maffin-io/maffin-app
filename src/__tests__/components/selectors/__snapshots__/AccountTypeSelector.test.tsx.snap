// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountTypeSelector passes data as expected 1`] = `
<div>
  <div
    data-testid="Selector"
  />
</div>
`;

exports[`AccountTypeSelector renders as expected 1`] = `
<div>
  <div
    data-testid="Selector"
  />
</div>
`;
